zenith/environment.graphqls

```diff
@@ -0,0 +1,119 @@
+# Draft for a Zenith extension to the Dagger API GraphQL schema
+
+extend type Query {
+    "The current environment"
+    here: Environment!
+
+    "Load a new environment"
+    environment(
+        "Source code of the environment"
+        source: Directory!,
+        "Working directory of the environment"
+        workdir: Directory!): Environment!
+    )
+}
+
+"API for your Dagger environment"
+type Environment {
+  "Command-line tools available in the environment"
+  tools: [Tool!]
+  "Lookup a command-line tool by name"
+  tool(name: String!): Tool!
+
+  "Artifacts available in the environment"
+  artifacts(filters: [String!]): [Artifact!]
+
+  """
+  Checks available in the environment
+  This could be unit tests, integration tests, linting, etc.
+  """
+  checks: [Checks!]
+  "Lookup a check by name"
+  check(name: String!): Check!
+
+  "Services available in the environment"
+  services: [Service!]
+  "Lookup a service by name"
+  service(name: String!): Service!
+  
+  "Interactive shells available in the environment"
+  shells: [Shell!]
+  "Lookup a shell by name"
+  shell(name: String!): Shell!
+
+  "The working directory for this environment"
+  workdir: Directory!
+
+  "Direct access to the environment as a data graph. Great for scripting and composition"
+  graph: EnvironmentGraph!
+
+  """
+  Extend this environment with the capabilities of other environments.
+  If no namespace is provided, environments are merged - this may cause naming conflicts.
+  """
+  withExtension(namespace: String, environment: Environment!): Environment!
```

Comment:
I kind of feel like this should happen in a static, declarative config file (dagger.json in the environment source, basically).

- I love that I can programmatically load and execute environments from anywhere - github, universe etc.
- But I also would like an easy way to add, remove and generally manage my extensions *without having to write code*. For example in a GUI. Also dependency management is a feature that users expect to have access to directly - to wrap it in code would probably be surprising and confusing to most.

Reply:
>  or "shadow code" with an increasingly complex DSL.

Agree we should avoid this at all costs. I'm okay with simple stuff like URIs (`local://`, `git://`, `universe://`, etc.), but we should cut it off at that level of complexity.

> But I also would like an easy way to add, remove and generally manage my extensions without having to write code. For example in a GUI. Also dependency management is a feature that users expect to have access to directly - to wrap it in code would probably be surprising and confusing to most.

I think this sort of approach would be viable:
1. The default is to put it in `dagger.json`, we support only simple URIs as mentioned above.
  * This would either be automatically included in your environment or we could require an explicit `WithExtensionsFromConfig` type of call. Automatic inclusion is probably my preference atm.
2. Very advanced use cases that need on-the-fly customization of the extension can still use the explicit `WithExtension` call.
  * This loses the simplicity of the static config, but it would only be needed for advanced users, so I think it's okay to make that tradeoff.
  * Also, if you ever need to do this, it means you are customizing an environment on the fly, which means you could instead just define a new environment that wraps the existing one with your customizations and then load that from your config instead.

tl;dr default to simple config, explicit `withExtension` is reserved for unblocking advanced use cases and has a clear pattern for migrating to the simple config.

Reply:
another possibility, perhaps too crazy: what if we assumed that with the advent of LLMs, software tools can be programmed to query and modify code as easily as a config file. And therefore we don’t need a “dumbed down” version of that code in a simplified json dsl: just regular code for everybody.

Maybe too radical?

Reply:
If we have static file + code, how do we deal with possible conflicts between them?

Reply:
> Can we have commands for modifying the dagger.json file?

Yeah we currently have `dagger project` for creating/reading dagger.json, but in the old incarnation of `dagger project` last year it also had commands for adding/updating dependencies.

Will take a closer look at the CLI PR soon, but at a first glance looks great!

> If we have static file + code, how do we deal with possible conflicts between them?

Same way that we deal with conflicts within the static file or conflicts within the code, error out.

I normally don't like two sources of truth, but as long as the static file is the "standard blessed way" and the code approach is reserved for our "Advanced Use Cases FAQ" type of documentation, I think it's totally fine. Express as much as you can in the file, if you are the 1% of users who need more flexibility, write some code too.

Reply:
You mean to save in `dagger.json` the environment schema when doing `dagger use github.com/vito/progrock@main `? I was thinking that would just save the URL in that file instead.

I get the performance benefits of having a "pre-loaded" schema though but it opens a can of worms in my mind so I need to catch up a bit more to have a more solid idea.

Reply:
One example for an advanced usage of `WithExtension` in code is to load extensions dynamically based on what's in the project. Like buildpacks.